[REST] basic input/output sanity check

jonasschnelli opened the pull request:
Mentioned by SergioDemianLerner at https://github.com/bitcoin/bitcoin/commit/f676c80f437e15ab0ee190e93baea733a88db0f2#commitcomment-8996114

Make sure that we not send back raw request parameters.

jonasschnelli committed:
[REST] basic input/output sanity check

Mentioned by SergioDemianLerner at https://github.com/bitcoin/bitcoin/commit/f676c80f437e15ab0ee190e93baea733a88db0f2#commitcomment-8996114

laanwj commented:
+1 for better input validation.

However I don't like the approach of using sanitize for input strings. It should either accept the input as-is or reject it and throw an error immediately. Don't try to proceed at all using 'sanitized' input.

Using sanitizestring for logged messages makes sense.

gavinandresen commented:
I agree with @laanwj -- throw an error if given unsanitary input, don't try to fix it.
(but sanitize if input is logged as part of the error handling, of course)

jonasschnelli commented:
Agreed. I'll update the PR to detect unsanitary inputs and throw exception in such cases.

jonasschnelli commented:
Updated.
Now it will detect unsafe inputs. Outputs still go though `SanitizeString()`.

laanwj commented on src/rest.cpp:
```diff
@@ -260,15 +277,15 @@ static bool rest_tx(AcceptedConnection* conn,
     const RetFormat rf = ParseDataFormat(params, strReq);
 
-    string hashStr = params[0];
+    CheckRequestString(params[0], IT_TXHASH);
```
Isn't this pre-check redundant? e.g., ParseHash will already fail on invalid hashes.

jonasschnelli committed:
[REST] basic input/output sanity check (overhaul)

- removed SanitizeString() because we now make sure that input parameters are correct.
- more precise parameter check including more details error string.

jonasschnelli commented:
@laanwj: Right. Is was redundant.

I just updated the whole PR.
I did also remove `SanitizeString()` from error replies because we are now sure that hashes and amounts are correct formatted, etc.
How does it look now?

Diapolo commented on src/rest.cpp:
```diff
@@ -52,6 +59,15 @@ static RestErr RESTERR(enum HTTPStatusCode status, string message)
     return re;
 }
 
+void CheckRequestString(const string inputString, enum InputType inputType = IT_UNDEF, size_t maxLength = 0)
```
Why are you not using ``const std::string& inputString``?

jonasschnelli replied:
Right.
The rest.cpp class was introduced with `using namespace std;` which is somehow inconsistent with other classes/files. This PR just follows the current rest.cpp implementation code style. But i'm pretty sure someone will open a PR to fix this. :)

Diapolo replied:
The important part wasn't the ``std::`` it was the ``&`` ;-).

jonasschnelli replied:
Ha. Yes. Right. will change that to pass-by-ref.

jonasschnelli replied:
> Why are you not using const std::string& inputString?

Fixed.

laanwj commented:
The redundant `CheckRequestString` check still seems to be there?

I'd personally have kept the SanitizeStrings on output just in case, as a kind of belt-and-suspenders, but no strong opinion there.

jonasschnelli commented:
@laanwj There is no more redundancy.  `CheckRequestString()` will ensure input correctness, `ParseHashStr()` will no longer check for correctness of the hashstr to parse.
I removed SanitizeStrings because it was no longer possible to output a non hex hash or unlimited-non-digi-value.

laanwj commented:
> ParseHashStr() will no longer check for correctness of the hashstr to parse.

To me it seems that was an excellent place to do the check. Where possible, combining checking input and parsing makes sense, as it reduces the scope for error.

jonasschnelli commented:
I'm also not happy with the solution. Let me try to optimize and shuffle around some things.
